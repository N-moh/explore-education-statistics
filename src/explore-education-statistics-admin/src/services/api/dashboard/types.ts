import { User } from '@admin/services/PrototypeLoginService';
import {
  DayMonthYearValues,
  IdLabelPair,
  UserContact,
} from '@admin/services/api/common/types/types';

export enum ReleaseApprovalStatus {
  Approved,
  ReadyToReview,
}

export interface ThemeAndTopics {
  title: string;
  id: string;
  topics: {
    title: string;
    id: string;
  }[];
}

export interface AdminDashboardRelease {
  id: string;
  status: ReleaseApprovalStatus;
  latestRelease: boolean;
  live: boolean;
  dateRangeLabel: string;
  timePeriodCoverage: IdLabelPair;
  leadStatistician: UserContact;
  lastEditedUser: User;
  lastEditedDateTime: string;
  scheduledPublishDate: DayMonthYearValues;
  nextReleaseExpectedDate: DayMonthYearValues;
}

export interface AdminDashboardPublication {
  id: string;
  title: string;
  methodology: IdLabelPair;
  releases: AdminDashboardRelease[];
}
