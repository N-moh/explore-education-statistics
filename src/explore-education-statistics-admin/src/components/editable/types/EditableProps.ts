export default interface EditableProps {
  canDelete?: boolean;
  editable?: boolean;
  onDelete?: () => void;
}
