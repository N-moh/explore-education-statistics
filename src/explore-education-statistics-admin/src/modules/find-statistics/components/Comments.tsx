import { useAuthContext } from '@admin/contexts/AuthContext';
import { useManageReleaseContext } from '@admin/pages/release/ManageReleaseContext';
import releaseContentCommentService, {
  AddExtendedComment,
  UpdateExtendedComment,
} from '@admin/services/releaseContentCommentService';
import { ExtendedComment } from '@admin/services/types/content';
import ButtonText from '@common/components/ButtonText';
import Details from '@common/components/Details';
import FormattedDate from '@common/components/FormattedDate';
import classNames from 'classnames';
import React, { useRef, useState } from 'react';
import styles from './Comments.module.scss';

export type CommentsChangeHandler = (
  blockId: string,
  comments: ExtendedComment[],
) => void;

interface Props {
  blockId: string;
  sectionId: string;
  comments: ExtendedComment[];
  canResolve?: boolean;
  canComment?: boolean;
  onChange: CommentsChangeHandler;
}

const Comments = ({
  blockId,
  sectionId,
  comments = [],
  onChange,
  canResolve = false,
  canComment = false,
}: Props) => {
  const ref = useRef<HTMLDivElement>(null);

  const [newComment, setNewComment] = useState<string>('');
  const [editableComment, setEditableComment] = useState<string>('');
  const [editableCommentText, setEditableCommentText] = useState<string>('');

  const { user } = useAuthContext();
  const { releaseId } = useManageReleaseContext();

  const addComment = (content: string) => {
    const additionalComment: AddExtendedComment = {
      content,
    };

    if (releaseId && sectionId) {
      releaseContentCommentService
        .addContentSectionComment(
          releaseId,
          sectionId,
          blockId,
          additionalComment,
        )
        .then(populatedComment => {
          let newComments = [populatedComment];

          if (comments) {
            newComments = [...newComments, ...comments];
          }

          onChange(blockId, newComments);
          setNewComment('');
        });
    }
  };

  const removeComment = (index: number) => {
    const commentId = comments[index].id;

    releaseContentCommentService
      .deleteContentSectionComment(commentId)
      .then(() => {
        const newComments = [...comments];
        newComments.splice(index, 1);

        onChange(blockId, newComments);
      });
  };

  const updateComment = (index: number, content: string) => {
    const editedComment: UpdateExtendedComment = {
      ...comments[index],
      content,
    };

    releaseContentCommentService
      .updateContentSectionComment(editedComment)
      .then(savedComment => {
        const newComments = [...comments];
        newComments[index] = savedComment;

        onChange(blockId, newComments);

        setEditableComment('');
        setEditableCommentText('');
      });
  };

  return (
    <>
      <div
        role="presentation"
        ref={ref}
        className={classNames('dfe-comment-block', [styles.addComment])}
      >
        <Details
          summary={`${canComment ? `Add / ` : ''}View comments (${
            comments.length
          })`}
          className="govuk-!-margin-bottom-1 govuk-body-s"
          onToggle={isOpen => {
            if (ref.current) {
              const section = document.getElementById(
                `content-section-${blockId}`,
              );
              if (isOpen) {
                ref.current.classList.add(styles.top);
                if (section) {
                  section.classList.add(styles.commentWhileCommenting);
                }
              } else {
                ref.current.classList.remove(styles.top);
                if (section) {
                  section.classList.remove(styles.commentWhileCommenting);
                }
              }
            }
          }}
        >
          {canComment && (
            <>
              <form>
                <textarea
                  name="comment"
                  id={`new_comment_${blockId}`}
                  value={newComment}
                  onChange={e => setNewComment(e.target.value)}
                />
                <button
                  type="button"
                  className="govuk-button"
                  disabled={newComment.length === 0}
                  onClick={() => {
                    addComment(newComment);
                  }}
                >
                  Submit
                </button>
              </form>
              <hr />
            </>
          )}
          <div className={styles.commentsContainer}>
            {comments.map(
              (
                { id, content, created, createdById, createdByName, updated },
                index,
              ) => (
                <div key={id}>
                  <h2 className="govuk-body-xs govuk-!-margin-0">
                    <strong>
                      {createdByName} <FormattedDate>{created}</FormattedDate>
                    </strong>
                  </h2>
                  {editableComment && editableComment === id ? (
                    <form>
                      <textarea
                        name="editComment"
                        id={`edit_comment_${id}`}
                        value={editableCommentText}
                        onChange={e => setEditableCommentText(e.target.value)}
                      />
                      <button
                        type="button"
                        className="govuk-button"
                        disabled={editableCommentText.length === 0}
                        onClick={() => {
                          updateComment(index, editableCommentText);
                        }}
                      >
                        Update
                      </button>
                    </form>
                  ) : (
                    <p className="govuk-body-xs govuk-!-margin-bottom-1">
                      {content}
                    </p>
                  )}
                  {canComment && (
                    <ButtonText
                      className="govuk-body-xs govuk-!-margin-right-3"
                      onClick={() => removeComment(index)}
                    >
                      Remove
                    </ButtonText>
                  )}
                  {canComment && user?.id === createdById && (
                    <>
                      {/* eslint-disable-next-line jsx-a11y/click-events-have-key-events */}
                      <ButtonText
                        className="govuk-body-xs govuk-!-margin-right-3"
                        onClick={() => {
                          setEditableCommentText(content);
                          return editableComment
                            ? setEditableComment('')
                            : setEditableComment(id);
                        }}
                      >
                        {editableComment && editableComment === id
                          ? 'Cancel'
                          : 'Edit'}
                      </ButtonText>
                    </>
                  )}
                  <hr />
                </div>
              ),
            )}
          </div>
        </Details>
      </div>
    </>
  );
};

export default Comments;
