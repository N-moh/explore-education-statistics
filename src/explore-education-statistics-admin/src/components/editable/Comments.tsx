import { useAuthContext } from '@admin/contexts/AuthContext';
import { useReleaseContext } from '@admin/pages/release/contexts/ReleaseContext';
import releaseContentCommentService, {
  AddComment,
  UpdateComment,
} from '@admin/services/releaseContentCommentService';
import { Comment } from '@admin/services/types/content';
import Button from '@common/components/Button';
import ButtonGroup from '@common/components/ButtonGroup';
import ButtonText from '@common/components/ButtonText';
import Details from '@common/components/Details';
import { Form } from '@common/components/form';
import FormFieldTextArea from '@common/components/form/FormFieldTextArea';
import FormattedDate from '@common/components/FormattedDate';
import Yup from '@common/validation/yup';
import { Formik } from 'formik';
import React, { useState } from 'react';
import orderBy from 'lodash/orderBy';
import styles from './Comments.module.scss';

interface FormValues {
  content: string;
}

export type CommentsChangeHandler = (
  blockId: string,
  comments: Comment[],
) => void;

interface Props {
  blockId: string;
  sectionId: string;
  comments: Comment[];
  canComment?: boolean;
  onChange: CommentsChangeHandler;
  onToggle?: (opened: boolean) => void;
}

const Comments = ({
  blockId,
  sectionId,
  comments = [],
  onChange,
  onToggle,
  canComment = true,
}: Props) => {
  const [editingComment, setEditingComment] = useState<Comment>();
  const { user } = useAuthContext();
  const { releaseId } = useReleaseContext();

  const addComment = (content: string) => {
    const additionalComment: AddComment = {
      content,
    };

    if (releaseId && sectionId) {
      releaseContentCommentService
        .addContentSectionComment(
          releaseId,
          sectionId,
          blockId,
          additionalComment,
        )
        .then(newComment => {
          onChange(blockId, [newComment, ...comments]);
        });
    }
  };

  const removeComment = (commentId: string) => {
    const index = comments.findIndex(comment => comment.id === commentId);
    releaseContentCommentService
      .deleteContentSectionComment(commentId)
      .then(() => {
        const newComments = [...comments];
        newComments.splice(index, 1);
        onChange(blockId, newComments);
      });
  };

  const updateComment = (commentId: string, content: string) => {
    const index = comments.findIndex(comment => comment.id === commentId);
    const editedComment: UpdateComment = {
      ...comments[index],
      content,
    };

    releaseContentCommentService
      .updateContentSectionComment(editedComment)
      .then(savedComment => {
        const newComments = [...comments];
        newComments[index] = savedComment;

        onChange(blockId, newComments);

        setEditingComment(undefined);
      });
  };

  const resolveComment = (commentId: string, resolved: boolean) => {
    const index = comments.findIndex(comment => comment.id === commentId);
    releaseContentCommentService
      .updateContentSectionComment(comments[index], resolved)
      .then(savedComment => {
        const newComments = [...comments];
        newComments[index] = savedComment;
        onChange(blockId, newComments);
      });
  };

  return (
    <div className={styles.container}>
      <Details
        onToggle={isOpen => onToggle && onToggle(isOpen)}
        className="govuk-!-margin-bottom-1 govuk-body-s"
        summary={`${canComment ? `Add / ` : ''}View comments (${
          comments.filter(comment => !comment.resolved).length
        } unresolved)`}
      >
        {canComment && (
          <Formik<FormValues>
            initialValues={{
              content: '',
            }}
            validationSchema={Yup.object({
              content: Yup.string().required('Enter a comment'),
            })}
            onSubmit={(values, { resetForm }) => {
              addComment(values.content);
              resetForm();
            }}
          >
            <Form id={`${blockId}-addCommentForm`} showErrorSummary={false}>
              <FormFieldTextArea<FormValues>
                label="Comment"
                name="content"
                rows={3}
              />

              <Button type="submit">Add comment</Button>
            </Form>
          </Formik>
        )}

        <ul className={styles.commentsContainer}>
          {orderBy(comments, ['created'], ['desc']).map(comment => {
            const { createdBy, resolvedBy } = comment;

            return comment.resolved ? (
              <li className="govuk-body-s">
                <p className="govuk-!-margin-0">
                  <strong>Comment resolved</strong>
                </p>
                <p className="govuk-!-margin-0">
                  <FormattedDate format="dd/MM/yy HH:mm">
                    {comment.resolved}
                  </FormattedDate>
                </p>
                <p className="govuk-!-margin-0">
                  {`by ${resolvedBy?.firstName} ${resolvedBy?.lastName}`}
                </p>
                <Details className="govuk-!-margin-0" summary="See comment">
                  <p className="govuk-!-margin-0">
                    <strong>
                      {`${createdBy.firstName} ${createdBy.lastName}`}
                    </strong>
                  </p>
                  <p className="govuk-!-margin-0">
                    {'Created: '}
                    <FormattedDate format="dd/MM/yy HH:mm">
                      {comment.created}
                    </FormattedDate>
                  </p>
                  {comment.updated && (
                    <p className="govuk-!-margin-0">
                      {'Updated: '}
                      <FormattedDate format="dd/MM/yy HH:mm">
                        {comment.updated}
                      </FormattedDate>
                    </p>
                  )}
                  <p className="govuk-!-margin-top-3">{comment.content}</p>
                </Details>
                <ButtonText
                  onClick={() => {
                    resolveComment(comment.id, false);
                  }}
                >
                  Unresolve
                </ButtonText>
                <hr />
              </li>
            ) : (
              <li key={comment.id} className="govuk-body-s">
                <p className="govuk-!-margin-0">
                  <strong>
                    {`${createdBy.firstName} ${createdBy.lastName}`}
                  </strong>
                </p>
                <p className="govuk-!-margin-0">
                  {'Created: '}
                  <FormattedDate format="dd/MM/yy HH:mm">
                    {comment.created}
                  </FormattedDate>
                </p>
                {comment.updated && (
                  <p className="govuk-!-margin-0">
                    {'Updated: '}
                    <FormattedDate format="dd/MM/yy HH:mm">
                      {comment.updated}
                    </FormattedDate>
                  </p>
                )}

                {editingComment?.id === comment.id ? (
                  <Formik<FormValues>
                    initialValues={{
                      content: editingComment.content,
                    }}
                    validationSchema={Yup.object({
                      content: Yup.string().required('Enter a comment'),
                    })}
                    onSubmit={values => {
                      updateComment(comment.id, values.content);
                    }}
                  >
                    <Form
                      id={`${blockId}-editCommentForm`}
                      showErrorSummary={false}
                    >
                      <FormFieldTextArea<FormValues>
                        label="Comment"
                        name="content"
                        id={`${blockId}-editCommentForm-editComment`}
                        rows={3}
                      />

                      <ButtonGroup>
                        <Button type="submit">Update</Button>
                        <ButtonText
                          onClick={() => {
                            setEditingComment(undefined);
                          }}
                        >
                          Cancel
                        </ButtonText>
                      </ButtonGroup>
                    </Form>
                  </Formik>
                ) : (
                  <>
                    <p className="govuk-!-margin-top-3">{comment.content}</p>
                    {canComment && (
                      <ButtonGroup>
                        {user?.id === createdBy.id && (
                          <>
                            <ButtonText
                              onClick={() => {
                                setEditingComment(comment);
                              }}
                            >
                              Edit
                            </ButtonText>
                            <ButtonText
                              onClick={() => {
                                removeComment(comment.id);
                              }}
                            >
                              Delete
                            </ButtonText>
                          </>
                        )}

                        <ButtonText
                          onClick={() => {
                            resolveComment(comment.id, true);
                          }}
                        >
                          Resolve
                        </ButtonText>
                      </ButtonGroup>
                    )}
                  </>
                )}

                <hr />
              </li>
            );
          })}
        </ul>
      </Details>
    </div>
  );
};

export default Comments;
