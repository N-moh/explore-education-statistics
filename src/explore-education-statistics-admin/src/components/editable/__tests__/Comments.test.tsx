import { AuthContext, User } from '@admin/contexts/AuthContext';
import { testRelease } from '@admin/pages/release/__data__/testRelease';
import { ReleaseContextProvider } from '@admin/pages/release/contexts/ReleaseContext';
import { GlobalPermissions } from '@admin/services/permissionService';
import _releaseContentCommentService from '@admin/services/releaseContentCommentService';
import { Comment } from '@admin/services/types/content';
import { render, screen, waitFor, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import noop from 'lodash/noop';
import React from 'react';
import Comments from '../Comments';

jest.mock('@admin/services/releaseContentCommentService');

const releaseContentCommentService = _releaseContentCommentService as jest.Mocked<
  typeof _releaseContentCommentService
>;

describe('Comments', () => {
  test('renders list of comments when comments are opened', () => {
    render(
      <ReleaseContextProvider release={testRelease}>
        <Comments
          blockId="test-block"
          sectionId="section-1"
          comments={[
            {
              id: 'comment-1',
              content: 'Test comment 1',
              createdBy: {
                id: 'user-1',
                email: 'test@test.com',
                firstName: 'John',
                lastName: 'Smith',
              },
              created: '2020-06-06T12:00:00',
            },
            {
              id: 'comment-2',
              content: 'Test comment 2',
              createdBy: {
                id: 'user-2',
                email: 'test2@test.com',
                firstName: 'Jane',
                lastName: 'Roberts',
              },
              created: '2020-06-06T11:00:00',
              updated: '2020-06-07T15:00:00',
            },
            {
              id: 'comment-3',
              content: 'Test comment 3',
              createdBy: {
                id: 'user-3',
                email: 'test3@test.com',
                firstName: 'John',
                lastName: 'Cale',
              },
              created: '2020-06-08T12:00:00',
              resolved: '2020-06-10T12:00:00',
              resolvedBy: {
                id: 'user-1',
                email: 'test@test.com',
                firstName: 'John',
                lastName: 'Smith',
              },
            },
          ]}
          onChange={noop}
        />
      </ReleaseContextProvider>,
    );

    userEvent.click(
      screen.getByRole('button', {
        name: 'Add / View comments (2 unresolved)',
      }),
    );

    const comments = screen.getAllByRole('listitem');

    expect(comments).toHaveLength(3);

    expect(comments[0]).toHaveTextContent('Comment resolved');
    expect(comments[0]).toHaveTextContent('On: 10/06/20 12:00');
    expect(comments[0]).toHaveTextContent('By: John Smith');
    expect(comments[0]).toHaveTextContent('See comment');
    expect(comments[0]).toHaveTextContent('Created: 08/06/20 12:00');
    expect(comments[0]).not.toHaveTextContent('Updated:');
    expect(comments[0]).toHaveTextContent('Test comment 3');

    expect(comments[1]).toHaveTextContent('John Smith');
    expect(comments[1]).toHaveTextContent('Created: 06/06/20 12:00');
    expect(comments[1]).toHaveTextContent('Test comment 1');
    expect(comments[1]).not.toHaveTextContent('Updated:');
    expect(comments[1]).not.toHaveTextContent('Comment resolved');

    expect(comments[2]).toHaveTextContent('Jane Roberts');
    expect(comments[2]).toHaveTextContent('Created: 06/06/20 11:00');
    expect(comments[2]).toHaveTextContent('Updated: 07/06/20 15:00');
    expect(comments[2]).toHaveTextContent('Test comment 2');
    expect(comments[2]).not.toHaveTextContent('Comment resolved');
  });

  test('adding comment calls the `onChange` handler with new comments list', async () => {
    const handleChange = jest.fn();

    render(
      <ReleaseContextProvider release={testRelease}>
        <Comments
          blockId="test-block"
          sectionId="section-1"
          comments={[
            {
              id: 'comment-1',
              content: 'Test comment 1',
              createdBy: {
                id: 'user-1',
                email: 'test@test.com',
                firstName: 'John',
                lastName: 'Smith',
              },
              created: '2020-06-06T12:00:00',
            },
          ]}
          onChange={handleChange}
        />
      </ReleaseContextProvider>,
    );

    await userEvent.type(
      screen.getByRole('textbox', {
        name: 'Comment',
        hidden: true,
      }),
      'New test comment',
    );

    releaseContentCommentService.addContentSectionComment.mockResolvedValue({
      id: 'comment-2',
      content: 'New test comment',
      createdBy: {
        id: 'user-2',
        email: 'test2@test.com',
        firstName: 'Bethany',
        lastName: 'Parker',
      },
      created: '2020-06-06T14:00:00',
    });

    userEvent.click(
      screen.getByRole('button', {
        name: 'Add comment',
        hidden: true,
      }),
    );

    await waitFor(() => {
      expect(handleChange).toHaveBeenCalledWith('test-block', [
        {
          id: 'comment-2',
          content: 'New test comment',
          createdBy: {
            id: 'user-2',
            email: 'test2@test.com',
            firstName: 'Bethany',
            lastName: 'Parker',
          },
          created: '2020-06-06T14:00:00',
        },
        {
          id: 'comment-1',
          content: 'Test comment 1',
          createdBy: {
            id: 'user-1',
            email: 'test@test.com',
            firstName: 'John',
            lastName: 'Smith',
          },
          created: '2020-06-06T12:00:00',
        },
      ] as Comment[]);
    });
  });

  describe('updating comment', () => {
    const testComments = [
      {
        id: 'comment-1',
        content: 'Test comment 1',
        createdBy: {
          id: 'user-1',
          email: 'test@test.com',
          firstName: 'John',
          lastName: 'Smith',
        },
        created: '2020-06-06T12:00:00',
      },
      {
        id: 'comment-2',
        content: 'Test comment 2',
        createdBy: {
          id: 'user-2',
          email: 'test2@test.com',
          firstName: 'Jane',
          lastName: 'Roberts',
        },
        created: '2020-06-06T11:00:00',
      },
    ];

    const testUser: User = {
      id: 'user-2',
      name: 'Jane Roberts',
      permissions: {} as GlobalPermissions,
    };

    test('can only update comment belonging to the current user', () => {
      render(
        <AuthContext.Provider
          value={{
            user: testUser,
          }}
        >
          <ReleaseContextProvider release={testRelease}>
            <Comments
              blockId="test-block"
              sectionId="section-1"
              comments={testComments}
              onChange={noop}
            />
          </ReleaseContextProvider>
        </AuthContext.Provider>,
      );

      const comments = screen.getAllByRole('listitem', {
        hidden: true,
      });

      expect(
        within(comments[0]).queryByRole('button', {
          name: 'Edit',
          hidden: true,
        }),
      ).toBeNull();

      expect(
        within(comments[1]).getByRole('button', {
          name: 'Edit',
          hidden: true,
        }),
      ).toBeInTheDocument();
    });

    test('clicking `Edit` button shows textbox and hides `Edit` button', () => {
      render(
        <AuthContext.Provider
          value={{
            user: testUser,
          }}
        >
          <ReleaseContextProvider release={testRelease}>
            <Comments
              blockId="test-block"
              sectionId="section-1"
              comments={testComments}
              onChange={noop}
            />
          </ReleaseContextProvider>
        </AuthContext.Provider>,
      );

      const comment = within(
        screen.getAllByRole('listitem', {
          hidden: true,
        })[1],
      );

      userEvent.click(
        comment.getByRole('button', {
          name: 'Edit',
          hidden: true,
        }),
      );

      expect(
        comment.getByRole('textbox', {
          name: 'Comment',
          hidden: true,
        }),
      ).toBeInTheDocument();

      expect(
        comment.queryByRole('button', {
          name: 'Edit',
          hidden: true,
        }),
      ).not.toBeInTheDocument();
    });

    test('clicking `Cancel` button hides the textbox again', () => {
      render(
        <AuthContext.Provider
          value={{
            user: testUser,
          }}
        >
          <ReleaseContextProvider release={testRelease}>
            <Comments
              blockId="test-block"
              sectionId="section-1"
              comments={testComments}
              onChange={noop}
            />
          </ReleaseContextProvider>
        </AuthContext.Provider>,
      );

      const comment = within(
        screen.getAllByRole('listitem', {
          hidden: true,
        })[1],
      );

      userEvent.click(
        comment.getByRole('button', {
          name: 'Edit',
          hidden: true,
        }),
      );

      userEvent.click(
        comment.getByRole('button', {
          name: 'Cancel',
          hidden: true,
        }),
      );

      expect(
        comment.queryByRole('textbox', {
          name: 'Comment',
          hidden: true,
        }),
      ).not.toBeInTheDocument();

      expect(
        comment.getByRole('button', {
          name: 'Edit',
          hidden: true,
        }),
      ).toBeInTheDocument();
    });

    test('updating comment calls the `onChange` handler with the updated comment', async () => {
      const handleChange = jest.fn();

      render(
        <AuthContext.Provider
          value={{
            user: testUser,
          }}
        >
          <ReleaseContextProvider release={testRelease}>
            <Comments
              blockId="test-block"
              sectionId="section-1"
              comments={testComments}
              onChange={handleChange}
            />
          </ReleaseContextProvider>
        </AuthContext.Provider>,
      );

      const comment = within(
        screen.getAllByRole('listitem', {
          hidden: true,
        })[1],
      );

      userEvent.click(
        comment.getByRole('button', {
          name: 'Edit',
          hidden: true,
        }),
      );

      await userEvent.type(
        comment.getByRole('textbox', {
          name: 'Comment',
          hidden: true,
        }),
        'Updated test comment',
      );

      releaseContentCommentService.updateContentSectionComment.mockResolvedValue(
        {
          id: 'comment-2',
          content: 'Updated test comment',
          createdBy: {
            id: 'user-2',
            email: 'test2@test.com',
            firstName: 'Jane',
            lastName: 'Roberts',
          },
          created: '2020-06-06T11:00:00',
          updated: '2020-06-06T15:00:00',
        },
      );

      userEvent.click(
        comment.getByRole('button', {
          name: 'Update',
          hidden: true,
        }),
      );

      await waitFor(() => {
        expect(handleChange).toHaveBeenCalledWith('test-block', [
          {
            id: 'comment-1',
            content: 'Test comment 1',
            createdBy: {
              id: 'user-1',
              email: 'test@test.com',
              firstName: 'John',
              lastName: 'Smith',
            },
            created: '2020-06-06T12:00:00',
          },
          {
            id: 'comment-2',
            content: 'Updated test comment',
            createdBy: {
              id: 'user-2',
              email: 'test2@test.com',
              firstName: 'Jane',
              lastName: 'Roberts',
            },
            created: '2020-06-06T11:00:00',
            updated: '2020-06-06T15:00:00',
          },
        ] as Comment[]);
      });
    });
  });

  describe('resolving comment', () => {
    const testUser: User = {
      id: 'user-1',
      name: 'Lou Reed',
      permissions: {} as GlobalPermissions,
    };

    test('clicking resolve button calls the `onChange` handler with updated comments', async () => {
      const testComments: Comment[] = [
        {
          id: 'comment-1',
          content: 'Test comment 1',
          createdBy: {
            id: 'user-1',
            email: 'test@test.com',
            firstName: 'Lou',
            lastName: 'Reed',
          },
          created: '2020-06-06T12:00:00',
        },
      ];

      const handleChange = jest.fn();

      render(
        <AuthContext.Provider
          value={{
            user: testUser,
          }}
        >
          <ReleaseContextProvider release={testRelease}>
            <Comments
              blockId="test-block"
              sectionId="section-1"
              comments={testComments}
              onChange={handleChange}
            />
          </ReleaseContextProvider>
        </AuthContext.Provider>,
      );

      const comments = screen.getAllByRole('listitem', {
        hidden: true,
      });

      expect(
        within(comments[0]).queryByRole('button', {
          name: 'Unresolve',
          hidden: true,
        }),
      ).not.toBeInTheDocument();

      expect(
        within(comments[0]).getByRole('button', {
          name: 'Resolve',
          hidden: true,
        }),
      ).toBeInTheDocument();

      const updateComment: Comment = {
        ...testComments[0],
        resolved: '2020-06-20T12:00:00',
        resolvedBy: {
          id: 'user-2',
          email: 'test@test.com',
          firstName: 'Jane',
          lastName: 'Roberts',
        },
        setResolved: true,
      };

      releaseContentCommentService.updateContentSectionComment.mockResolvedValue(
        updateComment,
      );

      userEvent.click(
        screen.getByRole('button', {
          name: 'Resolve',
          hidden: true,
        }),
      );

      await waitFor(() => {
        expect(handleChange).toHaveBeenCalledWith('test-block', [
          updateComment,
        ]);
      });
    });

    test('clicking unresolve button calls the `onChange` handler with the updated comments', async () => {
      const testComments: Comment[] = [
        {
          id: 'comment-2',
          content: 'Test comment 2',
          createdBy: {
            id: 'user-2',
            email: 'test2@test.com',
            firstName: 'Jane',
            lastName: 'Roberts',
          },
          created: '2020-06-06T11:00:00',
          updated: '2020-06-07T12:00:00',
          resolved: '2020-06-10T14:00:00',
          resolvedBy: {
            id: 'user-1',
            email: 'test@test.com',
            firstName: 'Lou',
            lastName: 'Reed',
          },
        },
      ];

      const handleChange = jest.fn();

      render(
        <AuthContext.Provider
          value={{
            user: testUser,
          }}
        >
          <ReleaseContextProvider release={testRelease}>
            <Comments
              blockId="test-block"
              sectionId="section-1"
              comments={testComments}
              onChange={handleChange}
            />
          </ReleaseContextProvider>
        </AuthContext.Provider>,
      );

      const comments = screen.getAllByRole('listitem', {
        hidden: true,
      });

      expect(
        within(comments[0]).queryByRole('button', {
          name: 'Resolve',
          hidden: true,
        }),
      ).not.toBeInTheDocument();

      expect(
        within(comments[0]).getByRole('button', {
          name: 'Unresolve',
          hidden: true,
        }),
      ).toBeInTheDocument();

      const updatedComment: Comment = {
        ...testComments[0],
        resolved: undefined,
        resolvedBy: undefined,
        setResolved: false,
      };

      releaseContentCommentService.updateContentSectionComment.mockResolvedValue(
        updatedComment,
      );

      userEvent.click(
        screen.getByRole('button', {
          name: 'Unresolve',
          hidden: true,
        }),
      );

      await waitFor(() => {
        expect(handleChange).toHaveBeenCalledWith('test-block', [
          updatedComment,
        ]);
      });
    });
  });

  describe('deleting comment', () => {
    const testComments: Comment[] = [
      {
        id: 'comment-1',
        content: 'Test comment 1',
        createdBy: {
          id: 'user-1',
          email: 'test@test.com',
          firstName: 'John',
          lastName: 'Smith',
        },
        created: '2020-06-06T12:00:00',
      },
      {
        id: 'comment-2',
        content: 'Test comment 2',
        createdBy: {
          id: 'user-2',
          email: 'test2@test.com',
          firstName: 'Jane',
          lastName: 'Roberts',
        },
        created: '2020-06-06T11:00:00',
      },
    ];

    const testUser: User = {
      id: 'user-2',
      name: 'Jane Roberts',
      permissions: {} as GlobalPermissions,
    };

    test('can only delete comment belonging to the current user', () => {
      render(
        <AuthContext.Provider
          value={{
            user: testUser,
          }}
        >
          <ReleaseContextProvider release={testRelease}>
            <Comments
              blockId="test-block"
              sectionId="section-1"
              comments={testComments}
              onChange={noop}
            />
          </ReleaseContextProvider>
        </AuthContext.Provider>,
      );

      const comments = screen.getAllByRole('listitem', {
        hidden: true,
      });

      expect(
        within(comments[0]).queryByRole('button', {
          name: 'Delete',
          hidden: true,
        }),
      ).toBeNull();

      expect(
        within(comments[1]).getByRole('button', {
          name: 'Delete',
          hidden: true,
        }),
      ).toBeInTheDocument();
    });

    test('deleting comment calls the `onChange` handler with it removed from comments', async () => {
      const handleChange = jest.fn();

      render(
        <AuthContext.Provider
          value={{
            user: testUser,
          }}
        >
          <ReleaseContextProvider release={testRelease}>
            <Comments
              blockId="test-block"
              sectionId="section-1"
              comments={testComments}
              onChange={handleChange}
            />
          </ReleaseContextProvider>
        </AuthContext.Provider>,
      );

      releaseContentCommentService.deleteContentSectionComment.mockResolvedValue();

      userEvent.click(
        screen.getByRole('button', {
          name: 'Delete',
          hidden: true,
        }),
      );

      await waitFor(() => {
        expect(handleChange).toHaveBeenCalledWith('test-block', [
          {
            id: 'comment-1',
            content: 'Test comment 1',
            createdBy: {
              id: 'user-1',
              email: 'test@test.com',
              firstName: 'John',
              lastName: 'Smith',
            },
            created: '2020-06-06T12:00:00',
          },
        ] as Comment[]);
      });
    });

    test('deleting comment calls the `onChange` handler with it removed from comments when selecting one of multiple comments', async () => {
      const handleChange = jest.fn();

      const testCommentsMultiple = [
        ...testComments,
        {
          id: 'comment-3',
          content: 'Test comment 3',
          createdBy: {
            id: 'user-2',
            email: 'test2@test.com',
            firstName: 'Jane',
            lastName: 'Roberts',
          },
          created: '2020-06-08T11:00:00',
        },
      ];

      render(
        <AuthContext.Provider
          value={{
            user: testUser,
          }}
        >
          <ReleaseContextProvider release={testRelease}>
            <Comments
              blockId="test-block"
              sectionId="section-1"
              comments={testCommentsMultiple}
              onChange={handleChange}
            />
          </ReleaseContextProvider>
        </AuthContext.Provider>,
      );

      releaseContentCommentService.deleteContentSectionComment.mockResolvedValue();

      userEvent.click(
        screen.getAllByRole('button', {
          name: 'Delete',
          hidden: true,
        })[0],
      );

      await waitFor(() => {
        expect(handleChange).toHaveBeenCalledWith('test-block', [
          {
            id: 'comment-1',
            content: 'Test comment 1',
            createdBy: {
              id: 'user-1',
              email: 'test@test.com',
              firstName: 'John',
              lastName: 'Smith',
            },
            created: '2020-06-06T12:00:00',
          },
          {
            id: 'comment-2',
            content: 'Test comment 2',
            createdBy: {
              id: 'user-2',
              email: 'test2@test.com',
              firstName: 'Jane',
              lastName: 'Roberts',
            },
            created: '2020-06-06T11:00:00',
          },
        ] as Comment[]);
      });
    });
  });
});
