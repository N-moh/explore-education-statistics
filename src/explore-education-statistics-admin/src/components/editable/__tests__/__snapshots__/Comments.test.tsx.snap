// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments renders list of comments when comments are opened 1`] = `
<ul
  class="commentsContainer"
>
  <li
    class="govuk-body-s"
  >
    <p
      class="govuk-!-margin-0"
    >
      <strong>
        John Smith
      </strong>
    </p>
    <p
      class="govuk-!-margin-0"
    >
      Created: 
      <time>
        06/06/20 12:00
      </time>
    </p>
    <p
      class="govuk-!-margin-top-3"
    >
      Test comment 1
    </p>
    <div
      class="group"
    >
      <button
        class="button"
        type="button"
      >
        Resolve
      </button>
    </div>
    <hr />
  </li>
  <li
    class="govuk-body-s"
  >
    <p
      class="govuk-!-margin-0"
    >
      <strong>
        Jane Roberts
      </strong>
    </p>
    <p
      class="govuk-!-margin-0"
    >
      Created: 
      <time>
        06/06/20 11:00
      </time>
    </p>
    <p
      class="govuk-!-margin-0"
    >
      Updated: 
      <time>
        07/06/20 15:00
      </time>
    </p>
    <p
      class="govuk-!-margin-top-3"
    >
      Test comment 2
    </p>
    <div
      class="group"
    >
      <button
        class="button"
        type="button"
      >
        Resolve
      </button>
    </div>
    <hr />
  </li>
</ul>
`;
